src/tsd.c

```diff
@@ -63,6 +63,48 @@ tsd_slow_update(tsd_t *tsd) {
 	}
 }
 
+static bool
+tsd_data_init(tsd_t *tsd) {
+	/*
+	 * We initialize the rtree context first (before the tcache), since the
+	 * tcache initialization depends on it.
+	 */
+	rtree_ctx_data_init(tsd_rtree_ctxp_get_unsafe(tsd));
+
+	if (tsd_tcache_enabled_data_init(tsd)) {
+		return true;
+	}
+	return false;
+}
+
+static void
+assert_tsd_data_cleanup_done(tsd_t *tsd) {
+	assert(!tsd_nominal(tsd));
+	assert(*tsd_arenap_get_unsafe(tsd) == NULL);
```

Comment:
Yeah it's possible to use x-macros, but might not be pretty. We need 2 columns: ```have_cleanup_val``` (some members like embedded tcache don't have a cleanup state), and ```cleanup_val```.  Also here we may want to assert something more intrusive, e.g. ```tcache.bins[0].avail == NULL```